src/merkle/merkle-objects.ts

```diff
@@ -0,0 +1,94 @@
+import CID from "cids";
+
+import { Doctype } from '@ceramicnetwork/common';
+import {
+  CompareFunction,
+  MergeFunction, MetadataFunction,
+  Node,
+  TreeMetadata,
+} from './merkle';
+
+import { logger } from '../logger';
+import { IpfsService } from '../services/ipfs-service';
+
+import bloom from 'bloomfilter.js';
+
+const BLOOM_FILTER_TYPE = "jsnpm_bloomfilter.js";
+
+export class Candidate {
+  public readonly cid: CID;
+  public readonly document: Doctype;
+  public readonly reqId: number;
+
+  constructor(cid: CID, reqId?: number, document?: Doctype) {
+    this.cid = cid;
+    this.reqId = reqId;
+    this.document = document
+  }
+
+  get docId(): string {
+    return this.document.id.baseID.toString()
+  }
+
+}
+
+/**
+ * Implements IPFS merge CIDs
+ */
+export class IpfsMerge implements MergeFunction<Candidate, TreeMetadata> {
+  private ipfsService: IpfsService;
+
+  constructor(ipfsService: IpfsService) {
+    this.ipfsService = ipfsService;
+  }
+
+  async merge(left: Node<Candidate>, right: Node<Candidate>, metadata: TreeMetadata | null): Promise<Node<Candidate>> {
+    const merged = [left.data.cid, right.data.cid];
+    if (metadata) {
+      const metadataCid = await this.ipfsService.storeRecord(metadata);
+      merged.push(metadataCid)
+    }
+
+    const mergedCid = await this.ipfsService.storeRecord(merged);
+    logger.debug('Merkle node ' + mergedCid + ' created.');
+    return new Node<Candidate>(new Candidate(mergedCid), left, right);
+  }
+}
+
+/**
+ * Implements IPFS merge CIDs
+ */
+export class IpfsLeafCompare implements CompareFunction<Candidate> {
+  compare(left: Node<Candidate>, right: Node<Candidate>): number {
+    return left.data.docId.localeCompare(right.data.docId);
+  }
+}
+
+/**
+ * Implements IPFS merge CIDs
+ */
+export class BloomMetadata implements MetadataFunction<Candidate, TreeMetadata> {
+  generateMetadata(leaves: Array<Node<Candidate>>): TreeMetadata {
+    const bloomFilterEntries = new Set<string>()
+    for (const node of leaves) {
+      const doc = node.data.document
+      bloomFilterEntries.add(`docid-${doc.id.baseID.toString()}`)
+      bloomFilterEntries.add(`schema-${doc.metadata.schema?.toString()}`)
+      bloomFilterEntries.add(`family-${doc.metadata.family}`) // todo should we use a special charactor of some kind to distingush the family "undefined" from no family?
```

Comment:
Currently if no schema or family is set, they still go into the bloom filter as `schema-undefined` or `family-undefined`.  Putting *something* in when nothing is provided makes sense to me cause I could imagine indexing services caring about docs with no schema defined at all, or no family defined at all.  In the case of schema, `schema-undefined` is fine to be the marker for "no schema was set" since "undefined" isn't a valid `CommitID` and thus can never be the actual value of the schema field.  For "family", however, someone could in theory set their "family" to the string "undefined", and we wouldn't be able to distinguish that from no family being set at all.  The only way I could imagine getting around that would be to have some special character that we forbid being present in the "family" field

Reply:
Actually nevermind, we can change the part that is hard-coded (`family-`) in a way that no user will be able to affect

Reply:
changed to `no-family` and `no-schema` as the markers in the bloom filter for when no family or no schema is provided